src/main/scala/viper/gobra/translator/encodings/interfaces/InterfaceEncoding.scala

```diff
@@ -349,4 +414,319 @@ class InterfaceEncoding extends LeafTypeEncoding {
+      receiver = p.receiver,
+      name = in.MethodProxy(p.superProxy.name, proofName(p.subProxy, p.superProxy))(p.info),
+      args = p.args,
+      results = p.results,
+      pres = Vector.empty,
+      posts = Vector.empty,
+      body = Some(body)
+    )(p.info))(ctx)
+
+    val pres = vItfFun.pres.map { pre =>
+      instantiateInterfaceSpecForProof(pre, vItfFun.formalArgs.toVector, p.receiver, p.args, p.superT)(p)(ctx)
+    }
+
+    val (pos, info, errT) = p.vprMeta
+    pureMethodDummy.map(res => Vector(res.copy(pres = pres)(pos, info, errT)))
+  }
+
+  /**
+    * method proof_T_I_M(x: T, args) returns (results)
+    *   requires PRE where PRE = [I_M.PRE][ this -> tuple2(x, TYPE(T)); I_P -> T_I_P_proof ]
+    *   ensures  POST where POST = [I_M.POST][ this -> tuple2(x, TYPE(T)); I_P -> T_I_P_proof ]
+    * {
+    *   [body]
+    * }
+    */
+  private def methodProof(p: in.MethodSubtypeProof)(ctx: Context): MemberWriter[Vector[vpr.Method]] = {
+
+    val itfSymb = ctx.lookup(p.superProxy).asInstanceOf[in.Method]
+    val vItfMeth = ctx.method.method(itfSymb)(ctx).res
+
+    val body = p.body.getOrElse {
+      val targets = p.results.map(out => in.LocalVar(out.id, out.typ)(out.info))
+      val call = in.MethodCall(targets, p.receiver, p.subProxy, p.args)(p.info)
+      in.Block(Vector.empty, Vector(call))(p.info)
+    }
+
+    val methodDummy = ctx.method.method(in.Method(
+      receiver = p.receiver,
+      name = in.MethodProxy(p.superProxy.name, proofName(p.subProxy, p.superProxy))(p.info),
+      args = p.args,
+      results = p.results,
+      pres = Vector.empty,
+      posts = Vector.empty,
+      body = Some(body)
+    )(p.info))(ctx)
+
+    val pres = vItfMeth.pres.map { exp =>
+      instantiateInterfaceSpecForProof(exp, vItfMeth.formalArgs.toVector, p.receiver, p.args, p.superT)(p)(ctx)
+    }
+    val posts = vItfMeth.posts.map { exp =>
+      instantiateInterfaceSpecForProof(exp, vItfMeth.formalArgs.toVector, p.receiver, p.args, p.superT)(p)(ctx)
+    }
+
+    val (pos, info, errT) = p.vprMeta
+    methodDummy.map(res => Vector(res.copy(pres = pres, posts = posts)(pos, info, errT)))
+  }
+
+
+  /** returns exp[ this -> tuple2(this, Type(this); I_P -> T_I_P_proof ] */
+  private def instantiateInterfaceSpecForProof(
+                         exp: vpr.Exp,
+                         formalsOfExp: Vector[vpr.LocalVarDecl],
+                         recv: in.Parameter.In,
+                         ins: Vector[in.Parameter.In],
+                         itfT: in.InterfaceT
+                       )(src: in.Node)(ctx: Context): vpr.Exp = {
+    val impl = recv.typ
+    val vRecvDecls = ctx.typeEncoding.variable(ctx)(recv)
+    val vRecv = vRecvDecls.localVar
+
+    val vArgDecls = ins map ctx.typeEncoding.variable(ctx)
+    val vArgs = vArgDecls map (_.localVar)
+
+    val itfFuncs = ctx.table.members(itfT).collect{ case x: in.MethodProxy if ctx.lookup(x).isInstanceOf[in.PureMethod] => x }
```

Comment:
You seem to replace pure interface method calls to their counterpart in the implementation. However, I cannot spot a similar treatment for mpredicates. Is this missing in the implementation? Doesn't this need to be done as well? And what does `I_P` and `T_I_P_proof` stand for? Doesn't the "P" stand for predicate?

Reply:
The mpredicate substitutions happens trivially, since both are literally the same predicate. I_P -> T_I_P_proof means that I replace a pure method of the interface with the proof that the implementation refined the interface.

Reply:
I tried to make it clearer in the code.